lnpbp-0021.md

```diff
@@ -1,213 +1,162 @@
-            format: StateFormat::CustomData(DataFormat::Unsigned(Bits::Bit64, 0, core::u64::MAX as u128)),
-            abi: none!()
-        },
-        OwnedRightsType::Ownership => StateSchema {
-            // This is unique token identifier, which is
-            // SHA256(SHA256(nft_source_state), issue_transition_id)
-            // convoluted to 32-bits with XOR operation and then XORed with
-            // salt value from state transition metadata.
-            // NB: It is unique inside single state transition only, not
-            // globally. For global unique id use non-convoluted hash value.
-            format: StateFormat::CustomData(DataFormat::Unsigned(Bits::Bit32, 0, core::u32::MAX as u128)),
-            abi: bmap! {
-                // Here we ensure that each unique state value is
-                // transferred once and only once (using "salt" value for
-                // collision resoultion)
-                AssignmentAction::Validate => Procedure::Embedded(StandardProcedure::IdentityTransfer)
-            }
-        },
-        OwnedRightsType::Renomination => StateSchema {
-            format: StateFormat::Declarative,
-            abi: none!()
-        }
-    },
-    public_right_types: none!(),
-}
+- [Abstract](#abstract)
+- [Background](#background)
+- [Motivation](#motivation)
+- [Design](#design)
+- [Specification](#specification)
+- [Compatibility](#compatibility)
+- [Rationale](#rationale)
+- [Reference implementation](#reference-implementation)
+- [Acknowledgements](#acknowledgements)
+- [References](#references)
+- [Copyright](#copyright)
+- [Test vectors](#test-vectors)
+
+
+## Abstract
+
+
+## Background
+
+
+## Motivation
+
+
+## Design
+
+- [x] Media as URI (which can be attachments distributed with Storm or usual URLs)
+- [x] Small media ("previews" if given together with large media)
+- [x] Fraction of assets
+- [x] Engravings
+- [x] Reserves
+- [x] Possible decentralized issue
+
+
+## Specification
+
+Interface specification is the following Contractum code:
+
+```haskell
+-- # Defining main data structures
+
+-- collectibles are usually scarse, so we limit their max number to 64k
+data ItemsCount :: U16
+
+-- each collectible item is unique and must have an id
+data TokenId :: U16
+
+-- Collectibles are owned in fractions of a single item; here the owned
+-- fraction means `1/F`. Ownership of the full item is specified `F=1`;
+-- half of an item as `F=2` etc.
+data OwnedFraction :: U64
+
+data TxOut :: txid [Byte ^ 32], vout U16
+
+-- allocation of a single token or its fraction to some transaction output
+data Allocation :: TokenId, OwnedFraction
+
+data Media ::
+    type MimeType,
+    data [Byte]
+
+data Attachment ::
+    type MimeType,
+    uri: Uri
+
+data POR :: -- proof of reserves
+    reserves TxOut,
+    proof [Byte] -- schema-specific proof 
+
+data Token ::
+    id TokenId,
+    name [Ascii ^ 1..40],
+    details [Unicode ^ 40..256]?,
+    media Media?,
+    attachment Attachment?,
```

Comment:
```suggestion
    preview Media?, -- always embedded preview media < 64kb
    media Attachment?, -- external media which is the main media for the token
    attachments { U8 ^ ..20 -> Attachment } -- auxillary attachments by type (up to 20 attachments)
```

Reply:
I don't suppose we could do something like 255 attachments? For example, if this is used for songs, instead of having an RGB21 for each song, we could have an RGB21 for each album, and albums rarely have more than 20 songs, and certainly fewer than 100. Not sure if there's a threshold where byte space efficiency makes sense to do 255 instead of say 100.

Also bear in mind, it's spelled `auxiliary`, not two ls, but an l and a i. With American pronunciation, it's said differently than it's spelled. I ordinarily wouldn't be so pedantic about spelling, but... Such a mistake might be harder to correct in the future, who knows.

Reply:
These attachments are not individual non-fungible tokens, but attachments to a single token, each of which have a unique MIME type. I.e. in case of ablum, the album is an contract under a scheme implementing RGB21, with each song being a `Token` here, which `media` is the audio recording, and `attachments` are lyrics, and maybe something else (cover image?). I doubt we need a lot of attachments, but I agree that 20 is arbitrary small amount.

Reply:
Right, I'm suggesting the limit should be higher so the token would be the album and not each individual song.

Reply:
Hm, and what would be contract in that case?

Making token the album will:
- disallow selling/owning songs by one
- leave them without names
- leave themselves without attachments (like do not providing lyrics)

At the same time it is not clear which advantages we would get...

Reply:
An album associates them. There is a way around this, and it's to have a `Parent?`, where there's a token for an album and each song, and each song specifies the RGB ID for the parent album, which would have no parent, only an artist, which is different since that's the issuer, though maybe it could point to an RGB22.

Reply:
Sorry, I am still not getting why we need this complexity if an RGB21 _contract_ can represent an album, and a token inside the contract a song.

Reply:
I see, so an album contract would have a token for each song... That's great, I like it! Each token will have a unique id, correct? To make them enumerable?

Reply:
Correct. That what I was trying to propose

Reply:
Sounds good. We can consider this point resolved, then.